uc.c

```diff
@@ -935,206 +918,72 @@ MemoryRegion *memory_mapping(struct uc_struct* uc, uint64_t address)
-
-    if (type & UC_HOOK_MEM_WRITE_PROT)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_WRITE_PROT, callback, user_data, hh);
-
-    if (type & UC_HOOK_MEM_FETCH_PROT)
-        ret = _hook_mem_invalid(uc, UC_HOOK_MEM_FETCH_PROT, callback, user_data, hh);
+    struct hook *hook = calloc(1, sizeof(struct hook));
+    if (hook == NULL) {
+        return UC_ERR_NOMEM;
+    }
+    hook->type = type;
+    hook->callback = callback;
+    hook->user_data = user_data;
+    *hh = (uc_hook)hook;
 
-    switch(type) {
-        default:
-            break;
-        case UC_HOOK_INTR:
-            ret = _hook_intr(uc, callback, user_data, hh);
-            break;
-        case UC_HOOK_INSN:
-            id = va_arg(valist, int);
-            ret = _hook_insn(uc, id, callback, user_data, hh);
-            break;
-        case UC_HOOK_CODE:
-            begin = va_arg(valist, uint64_t);
-            end = va_arg(valist, uint64_t);
-            ret = _hook_code(uc, UC_HOOK_CODE, begin, end, callback, user_data, hh);
-            break;
-        case UC_HOOK_BLOCK:
-            begin = va_arg(valist, uint64_t);
-            end = va_arg(valist, uint64_t);
-            ret = _hook_code(uc, UC_HOOK_BLOCK, begin, end, callback, user_data, hh);
-            break;
-        case UC_HOOK_MEM_READ:
-            begin = va_arg(valist, uint64_t);
-            end = va_arg(valist, uint64_t);
-            ret = _hook_mem_access(uc, UC_HOOK_MEM_READ, begin, end, callback, user_data, hh);
-            break;
-        case UC_HOOK_MEM_WRITE:
-            begin = va_arg(valist, uint64_t);
-            end = va_arg(valist, uint64_t);
-            ret = _hook_mem_access(uc, UC_HOOK_MEM_WRITE, begin, end, callback, user_data, hh);
-            break;
-        case UC_HOOK_MEM_READ | UC_HOOK_MEM_WRITE:
-            begin = va_arg(valist, uint64_t);
-            end = va_arg(valist, uint64_t);
-            ret = _hook_mem_access(uc, UC_HOOK_MEM_READ | UC_HOOK_MEM_WRITE, begin, end, callback, user_data, hh);
-            break;
+    // everybody but HOOK_INSN gets begin/end, so exit early here.
+    if (type & UC_HOOK_INSN) {
+        hook->insn = va_arg(valist, int);
+        if (list_append(&uc->hook[UC_HOOK_INSN_IDX], hook) == NULL) {
+            return UC_ERR_NOMEM;
+        }
+        return UC_ERR_OK;
     }
 
+    hook->begin = va_arg(valist, uint64_t);
+    hook->end = va_arg(valist, uint64_t);
     va_end(valist);
 
+    int i = 0;
+    while ((type >> i) > 0) {
+        if ((type >> i) & 1) {
+            // TODO: invalid hook error?
+            if (i < UC_HOOK_MAX) {
+                if (list_append(&uc->hook[i], hook) == NULL) {
+                    return UC_ERR_NOMEM;
+                }
+            }
+        }
+        i++;
+    }
+
     return ret;
 }
 
 UNICORN_EXPORT
 uc_err uc_hook_del(uc_engine *uc, uc_hook hh)
 {
-    return hook_del(uc, hh);
+    int i;
+    for (i = 0; i < UC_HOOK_MAX; i++) {
+        list_remove(&uc->hook[i], (void *)hh);
```

Comment:
this does not look good, because this loop always go thru all the entries even list_remove() already removed the hook at some point.

also, the hook is not freed anywhere, so memory is leaking.